include/graphblas/reference/blas3.hpp

```diff
@@ -918,142 +946,108 @@ namespace grb {
+#endif
+				return ILLEGAL;
+			}
 
 			if( m != m_A || m != m_B || n != n_A || n != n_B ) {
+#ifdef _DEBUG
+				std::cerr << "grb::eWiseApply: dimensions of input matrices do not match\n";
+#endif
 				return MISMATCH;
 			}
 
+			if( getID(A) == getID(C) || getID(B) == getID(C) ) {
+#ifdef _DEBUG
+				std::cerr << "grb::eWiseApply: The output matrix can not simultaneously be "
+					<< "one of the input matrices\n";
+#endif
+			}
+
 			const auto &A_raw = !trans_left ?
 				internal::getCRS( A ) :
 				internal::getCCS( A );
 			const auto &B_raw = !trans_right ?
 				internal::getCRS( B ) :
 				internal::getCCS( B );
-			auto &C_raw = internal::getCRS( C );
+			auto &CRS_raw = internal::getCRS( C );
 			auto &CCS_raw = internal::getCCS( C );
-
-#ifdef _DEBUG
-			std::cout << "\t\t A offset array = { ";
-			for( size_t i = 0; i <= m_A; ++i ) {
-				std::cout << A_raw.col_start[ i ] << " ";
-			}
-			std::cout << "}\n";
-			for( size_t i = 0; i < m_A; ++i ) {
-				for( size_t k = A_raw.col_start[ i ]; k < A_raw.col_start[ i + 1 ]; ++k ) {
-					std::cout << "\t\t ( " << i << ", " << A_raw.row_index[ k ] << " ) = "
-						<< A_raw.getPrintValue( k ) << "\n";
-				}
-			}
-			std::cout << "\t\t B offset array = { ";
-			for( size_t j = 0; j <= m_B; ++j ) {
-				std::cout << B_raw.col_start[ j ] << " ";
-			}
-			std::cout << "}\n";
-			for( size_t j = 0; j < m_B; ++j ) {
-				for( size_t k = B_raw.col_start[ j ]; k < B_raw.col_start[ j + 1 ]; ++k ) {
-					std::cout << "\t\t ( " << B_raw.row_index[ k ] << ", " << j << " ) = "
-						<< B_raw.getPrintValue( k ) << "\n";
-				}
-			}
-#endif
+			const auto dummy_identity = identities::zero< OutputType >::value();
 
 			// retrieve buffers
-			char * arr1, * arr2, * arr3, * buf1, * buf2, * buf3;
-			arr1 = arr2 = buf1 = buf2 = nullptr;
+			char * arr1 = nullptr, * arr3 = nullptr, * buf1 = nullptr, * buf3 = nullptr;
 			InputType1 * vbuf1 = nullptr;
-			InputType2 * vbuf2 = nullptr;
 			OutputType * valbuf = nullptr;
 			internal::getMatrixBuffers( arr1, buf1, vbuf1, 1, A );
-			internal::getMatrixBuffers( arr2, buf2, vbuf2, 1, B );
 			internal::getMatrixBuffers( arr3, buf3, valbuf, 1, C );
 			// end buffer retrieval
 
-			// initialisations
-			internal::Coordinates< reference > coors1, coors2;
+			// initialisations of the coordinates
+			Coordinates<reference> coors1;
 			coors1.set( arr1, false, buf1, n );
-			coors2.set( arr2, false, buf2, n );
+			// end initialisations of the coordinates
+
+			if( !crs_only ) {
 #ifdef _H_GRB_REFERENCE_OMP_BLAS3
-			#pragma omp parallel
-			{
-				size_t start, end;
-				config::OMP::localRange( start, end, 0, n + 1 );
-#else
-				const size_t start = 0;
-				const size_t end = n + 1;
+				#pragma omp parallel for simd default(none) \
+					shared(CCS_raw) firstprivate(n)
 #endif
-				for( size_t j = start; j < end; ++j ) {
+				for( size_t j = 0; j <= n; ++j ) {
 					CCS_raw.col_start[ j ] = 0;
 				}
-#ifdef _H_GRB_REFERENCE_OMP_BLAS3
 			}
-#endif
```

Comment:
btw if the openmp default is static (initially I thought dynamic, but in hindsight and a little bit later I'm not sure), please still make it explicitly static to avoid ambiguity. Also note that by using `OMP::localRange`, ALP also controls exactly how the static schedule is derived, which I find preferable. Localrange also takes into account false sharing issues.